src/main/scala/mesosphere/marathon/core/launcher/impl/InstanceOpFactoryImpl.scala

```diff
@@ -125,7 +126,13 @@ class InstanceOpFactoryImpl(
         config.defaultAcceptedResourceRolesSet, config, schedulerPlugins, localRegion)
     matchResponse match {
       case matches: ResourceMatchResponse.Match =>
-        val taskId = Task.Id.forInstanceId(scheduledInstance.instanceId)
+
+        // this will either create a new TaskId with an adjusted incarnation counter
+        // based on an existing taskId, or create a new TaskId with incarnation count 1.
+        val maybeExistingTaskId = scheduledInstance.tasksMap.values.headOption.map(_.taskId)
+        val taskId = maybeExistingTaskId.map(Task.Id.withIncarnationCount).getOrElse {
+          TaskIdWithIncarnation(scheduledInstance.instanceId, None, 1L)
+        }
```

Comment:
We probably want to move this logic to `Task.Id.forInstanceId`.

Reply:
I am not exactly sure how would I do that... Do you mind proposing something? The code in Task class is pretty new to me 🙈

Reply:
Hm, it feels like we should move this into `scheduledInstance.provisioned`. In any case I'd do 

```scala
 val taskId = maybeExistingTaskId.map(Task.Id.withIncarnationCount)
     .getOrElse(Task.Id.forInstanceId(scheduledInstance.instanceId))
```

Reply:
➕ 1